app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java

```diff
@@ -0,0 +1,183 @@
+
+    @StringDef({
+            ProviderType.LOCATION_PROVIDER_GPS,
+            ProviderType.LOCATION_PROVIDER_NETWORK,
+            ProviderType.LOCATION_PROVIDER_PASSIVE
+    })
+    @Retention(RetentionPolicy.SOURCE)
+    public @interface ProviderType {
+        String LOCATION_PROVIDER_GPS = LocationManager.GPS_PROVIDER;
+        String LOCATION_PROVIDER_NETWORK = LocationManager.NETWORK_PROVIDER;
+        String LOCATION_PROVIDER_PASSIVE = LocationManager.PASSIVE_PROVIDER;
+    }
+
+
+    private static void checkForLocationPermission(Context context) {
+        int fineLocationPermissionStatus = ActivityCompat.checkSelfPermission(context,
+                Manifest.permission.ACCESS_FINE_LOCATION);
+
+        if (fineLocationPermissionStatus != PackageManager.PERMISSION_GRANTED) {
+            throw new SecurityException("No permission granted for location fetching");
+        }
+    }
+
+
+    private Context mContext;
+
+    private LocationManager mLocationManager;
+
+    /**
+     * Underlying "native" listener
+     */
+    private LocationListener mLocationListener;
+
+    /**
+     * Location updates notifier
+     */
+    private LocationFetcherListener mLocationFetcherListener;
+
+    private Collection<String> mProviders;
+    private boolean mIsRegistered;
+
+    public LocationFetcher(Context context, LocationManager locationManager,
+                           LocationFetcherListener locationFetcherListener) {
+        mContext = context;
+        mLocationManager = locationManager;
+        mLocationFetcherListener = locationFetcherListener;
+        mLocationListener = new MyLocationListener();
+
+        mProviders = new ArrayList<>();
+        mIsRegistered = false;
+    }
+
+    /**
+     * Adds provider to be used when registering the fetcher
+     *
+     * @param locationProvider - type of provider to use for location updates
+     */
+    public void addProvider(@ProviderType String locationProvider) {
+        if (locationProvider == null || locationProvider.isEmpty()) {
+            throw new IllegalArgumentException("location provider cannot be null or empty");
+        }
+
+        mProviders.add(locationProvider);
+    }
+
+
+    /**
+     * Registers fetcher for location updates
+     *
+     * @param minSamplingFrequencyMs         - minimum time between location samples,  in milliseconds
+     * @param minDistanceDeltaSamplingMeters - minimum distance between location samples, in meters
+     */
+    public void registerForUpdates(long minSamplingFrequencyMs,
+                                   float minDistanceDeltaSamplingMeters) {
+        if (minSamplingFrequencyMs < 0) {
+            throw new IllegalArgumentException("minSamplingFrequencyMs cannot be negative");
+        }
+
+        if (minDistanceDeltaSamplingMeters < 0) {
+            throw new IllegalArgumentException(
+                    "minDistanceDeltaSamplingMeters cannot be a negative number");
+        }
+
+        if (mIsRegistered) {
```

Comment:
It should use another LocationFetcher.
It keeps a fetcher from getting new type of sources after registration.
I've also added a conditional at addProvider to help with that.